Fix missing shared addresses on restore

jrainville opened the pull request:
Not ready for review yet.

The test passes, but it requires the same ugly hack @mprakhov used to wait for the owner to receive the request.
Please send help

jrainville commented on protocol/communities_messenger_token_permissions_test.go:
```diff
@@ -2054,6 +2054,59 @@ func (s *MessengerCommunitiesTokenPermissionsSuite) TestResendEncryptionKeyOnBac
+	err = s.bob.HandleBackup(s.bob.buildMessageState(), backupMessage, nil)
+	s.Require().NoError(err)
+
+	// FIXME this is necessary to have the owner receive the request and there is no response
+	_, err = WaitOnMessengerResponse(
+		s.owner,
+		func(r *MessengerResponse) bool {
+			if len(r.requestsToJoinCommunity) > 0 && r.requestsToJoinCommunity[community.IDString()] != nil {
+				return true
+			}
+			return false
+		},
+		"request to join not received",
+	)
```
@mprakhov can you give me a hint how to solve this? `handleCommunitySharedAddressesRequest` doesn't have any response, so I don't know what to wait for. Can we hook to the message ending and wait for the response to have been sent?

mprakhov replied:
The easiest way will be to put the address sending into `Subscription` and in the test subscribe to this event
In other case, you'll have to add some interface into `handleRetrievedMessages`, which operates with `CurrentMessageState` (like `saveDataAndPrepareResponse`) to mock it and read the msg, which was processed

jrainville replied:
I just tried using Subscriptions, but it doesn't work for me, because I need to trigger `RetrieveAll` for the Owner.

I ended up using `BackupHandled` on the Response to say when the owner handled the response.

It's not super clean, but it works. If you prefer, I can introduce a new property on the MessengerResponse, but it would only be for tests.

mprakhov replied:
It will work, but it will require some extra work:
1. Set `Subscription` and implement resending address via subscription (example: `resendMemberAddress`)
2. In handle - send msg via Subscription (like we send `CommunityDescription` updates, see `saveAndPublish`)
3. Update `MessengerSignalsHandlerMock` with an added subscription method (`resendMemberAddress`)
4. Subscribe to your msg in `MessengerSignalsHandlerMock` 
5. In the owner's `WaitOnMessengerResponse` you check if your subscription method was called (`resendMemberAddress`)

jrainville replied:
Do you think I should redo my code/test to use that or using the MessengerResponse is good enough?

mprakhov replied:
For me - `BackupHandled` is ok for now, but we can create a tech debt ticket to fix it in the future. Up to you, this is definitely not a PR blocker

jrainville replied:
Opened here: https://github.com/status-im/status-go/issues/5326

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (28)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_multiplication_x: | f14b2f8a | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5314/1/) | 2024-06-07 18:54:41 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5314/1//consoleText) |
| | | | | | | |
| :heavy_check_mark: | f14b2f8aa59c9ff6371ee583b9a8e43b7a596e80 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5314/1/) | 2024-06-07 18:57:40 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240607-185332-f14b2f8aa59c9ff6371ee583b9a8e43b7a596e80-pr5314.zip) |
| :heavy_check_mark: | f14b2f8aa59c9ff6371ee583b9a8e43b7a596e80 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5314/1/) | 2024-06-07 18:57:58 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240607-185328-f14b2f8aa59c9ff6371ee583b9a8e43b7a596e80-pr5314.zip) |
| :heavy_check_mark: | f14b2f8aa59c9ff6371ee583b9a8e43b7a596e80 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5314/1/) | 2024-06-07 18:59:19 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240607-185328-f14b2f8aa59c9ff6371ee583b9a8e43b7a596e80-pr5314.aar) |
| | | | | | | |
| :x: | 30506b77dbc6f6287f10931fae76e0bbf51bde59 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5314/2/) | 2024-06-10 19:05:33 | ~29 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5314/2/consoleText) |
| :x: | 30506b77dbc6f6287f10931fae76e0bbf51bde59 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5314/2/) | 2024-06-10 19:05:33 | ~29 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5314/2/consoleText) |
| :x: | 30506b77dbc6f6287f10931fae76e0bbf51bde59 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5314/2/) | 2024-06-10 19:05:38 | ~32 sec | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5314/2/consoleText) |
| | | | | | | |
| :heavy_multiplication_x: | 30506b77 | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5314/2/) | 2024-06-10 19:07:03 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5314/2//consoleText) |
| | | | | | | |
| :heavy_multiplication_x: | 9a5c9fef | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5314/3/) | 2024-06-10 19:09:04 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5314/3//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 9a5c9fef4a7c74a71fe6ba18230851fb8063f89b | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5314/3/) | 2024-06-10 19:10:07 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240610-190738-9a5c9fef4a7c74a71fe6ba18230851fb8063f89b-pr5314.zip) |
| :heavy_check_mark: | 9a5c9fef4a7c74a71fe6ba18230851fb8063f89b | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5314/3/) | 2024-06-10 19:10:17 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240610-190738-9a5c9fef4a7c74a71fe6ba18230851fb8063f89b-pr5314.aar) |
| :heavy_check_mark: | 9a5c9fef4a7c74a71fe6ba18230851fb8063f89b | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5314/3/) | 2024-06-10 19:13:03 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240610-190738-9a5c9fef4a7c74a71fe6ba18230851fb8063f89b-pr5314.zip) |
| | | | | | | |
| :heavy_check_mark: | d914bd6ca3983f1a13987434e37b998d8fc78eb6 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5314/4/) | 2024-06-10 19:21:56 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240610-191933-d914bd6ca3983f1a13987434e37b998d8fc78eb6-pr5314.aar) |
| :heavy_check_mark: | d914bd6ca3983f1a13987434e37b998d8fc78eb6 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5314/4/) | 2024-06-10 19:22:41 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240610-191937-d914bd6ca3983f1a13987434e37b998d8fc78eb6-pr5314.zip) |
| :heavy_check_mark: | d914bd6ca3983f1a13987434e37b998d8fc78eb6 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5314/4/) | 2024-06-10 19:23:23 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240610-191937-d914bd6ca3983f1a13987434e37b998d8fc78eb6-pr5314.zip) |
| | | | | | | |
| :heavy_check_mark: | d914bd6c | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5314/4/) | 2024-06-10 20:03:12 | ~43 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5314/4//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 976aef7d041d32b124e3ce3e587d93b41be617b7 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5314/5/) | 2024-06-11 18:39:51 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240611-183711-976aef7d041d32b124e3ce3e587d93b41be617b7-pr5314.zip) |
| :heavy_check_mark: | 976aef7d041d32b124e3ce3e587d93b41be617b7 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5314/5/) | 2024-06-11 18:39:58 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240611-183711-976aef7d041d32b124e3ce3e587d93b41be617b7-pr5314.aar) |
| :heavy_check_mark: | 976aef7d041d32b124e3ce3e587d93b41be617b7 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5314/5/) | 2024-06-11 18:40:22 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240611-183711-976aef7d041d32b124e3ce3e587d93b41be617b7-pr5314.zip) |
| | | | | | | |
| :heavy_multiplication_x: | 976aef7d | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5314/5/) | 2024-06-11 19:18:10 | ~40 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5314/5//consoleText) |
| | | | | | | |
| :heavy_check_mark: | f8b3c85bb686ee9d980054e2b9b2dad44bc5833a | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5314/6/) | 2024-06-12 18:20:08 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240612-181733-f8b3c85bb686ee9d980054e2b9b2dad44bc5833a-pr5314.zip) |
| :heavy_check_mark: | f8b3c85bb686ee9d980054e2b9b2dad44bc5833a | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5314/6/) | 2024-06-12 18:20:16 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240612-181733-f8b3c85bb686ee9d980054e2b9b2dad44bc5833a-pr5314.aar) |
| :heavy_check_mark: | f8b3c85bb686ee9d980054e2b9b2dad44bc5833a | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5314/6/) | 2024-06-12 18:20:17 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240612-181733-f8b3c85bb686ee9d980054e2b9b2dad44bc5833a-pr5314.zip) |
| | | | | | | |
| :heavy_check_mark: | f8b3c85b | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5314/6/) | 2024-06-12 18:58:50 | ~41 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5314/6//consoleText) |
| | | | | | | |
| :heavy_multiplication_x: | d6f0e0ed | [#7](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5314/7/) | 2024-06-14 14:33:29 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5314/7//consoleText) |
| | | | | | | |
| :heavy_check_mark: | d6f0e0ed4f0595f26bd1c7b15998a6f7a0a58473 | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5314/7/) | 2024-06-14 14:34:51 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240614-143219-d6f0e0ed4f0595f26bd1c7b15998a6f7a0a58473-pr5314.aar) |
| :heavy_check_mark: | d6f0e0ed4f0595f26bd1c7b15998a6f7a0a58473 | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5314/7/) | 2024-06-14 14:35:02 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240614-143219-d6f0e0ed4f0595f26bd1c7b15998a6f7a0a58473-pr5314.zip) |
| :heavy_check_mark: | d6f0e0ed4f0595f26bd1c7b15998a6f7a0a58473 | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5314/7/) | 2024-06-14 14:36:19 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240614-143219-d6f0e0ed4f0595f26bd1c7b15998a6f7a0a58473-pr5314.zip) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 5a7a8a3a2d847a51a8424deba35e1e6155ace4c3 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5314/8/) | 2024-06-14 14:45:52 | ~1 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240614-144414-5a7a8a3a2d847a51a8424deba35e1e6155ace4c3-pr5314.aar) |
| :heavy_check_mark: | 5a7a8a3a2d847a51a8424deba35e1e6155ace4c3 | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5314/8/) | 2024-06-14 14:46:31 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240614-144414-5a7a8a3a2d847a51a8424deba35e1e6155ace4c3-pr5314.zip) |
| :heavy_check_mark: | 5a7a8a3a2d847a51a8424deba35e1e6155ace4c3 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5314/8/) | 2024-06-14 14:46:49 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240614-144414-5a7a8a3a2d847a51a8424deba35e1e6155ace4c3-pr5314.zip) |
| | | | | | | |
| :heavy_check_mark: | 5a7a8a3a | [#8](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5314/8/) | 2024-06-14 15:26:21 | ~42 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5314/8//consoleText) |

jrainville committed:
fix_: send request for shared addresses when restoring a community

Needed for https://github.com/status-im/status-desktop/issues/14289

When someone restores a community that they joined, it now sends a message to the control node to get their shared addresses back.

It's lighter that backing up the addresses all the time on waku and should only be needed once.

osmaczko commented on protocol/messenger_backup_handler.go:
```diff
@@ -304,22 +304,27 @@ func (m *Messenger) handleWatchOnlyAccount(message *protobuf.SyncAccount) error
 
 func (m *Messenger) handleSyncedCommunities(state *ReceivedMessageState, message *protobuf.Backup) []error {
 	var errors []error
+	handledCommunities := make(map[string]struct{})
```
I'm not sure I understand why we need that. Does the backup message contain duplicates? 🤔 If that's the case, then I believe a much cleaner solution would be to filter out duplicate communities before the loop. This way, there would be no need for the`alreadyHandled` parameter, which is passed to `requestCommunityKeysAndSharedAddresses`, as it doesn't make sense in the context of that function.

jrainville replied:
I don't know that code well enough, so I didn't want to introduce any major change.

As I could see, I was calling `request` more than just once. I think it's because when we fetch we get multiple copies of the backup?

I didn't wrap the whole loop in the `alreadyHandled` condition, because if my assumption that we can get multiple copies of the community is right, I don't want to miss  the latest one if we handled an old one.

But maybe that whole code is useless. I don't mind removing it. I just found it weird that I was calling the request multiple times.

osmaczko replied:
https://github.com/status-im/status-go/commit/1d6789150414c1344f319213412dfb68ecd5845c

I created a diff to explain what I wanted to say (code says more than thousand words :sweat_smile: )

jrainville replied:
Very clean! I did think of doing that, but I got lazy I guess :sweat_smile: I'll add your commit to my PR

osmaczko commented on protocol/messenger_backup_handler.go:
```diff
@@ -304,22 +304,27 @@ func (m *Messenger) handleWatchOnlyAccount(message *protobuf.SyncAccount) error
 
-		err = m.requestCommunityKeys(state, syncCommunity)
+		_, alreadyHandled := handledCommunities[idString]
+		err = m.requestCommunityKeysAndSharedAddresses(state, syncCommunity, alreadyHandled)
```
Do we want to request shared addresses each time we handle backup, even outdated one?

jrainville replied:
Don't we handle backup only once when we restore the account?

osmaczko replied:
As far as I can see, we do enable handing backup messages when a messenger is created:

```golang
if s.config.ProcessBackedupMessages {
    s.messenger.EnableBackedupMessagesProcessing()
}
```

I assume, this is only the case when we are recovering account from seedphrase, but even though, it means backup messages processing is enabled for that login session. 

@saledjenic can you please check my reasoning above?

If the above is true, then if we receive N backup messages from store nodes, N requests for shared addresses will be sent, which is not ideal.

jrainville replied:
Right, if that's really the case, I'll indeed need to update the code, though we'll need to update the part below too as we request keys for communities too

osmaczko replied:
Oh, true, I thought we are checking if we have missing keys before we ask but seems we don't :thinking:

IMO, asking for keys and shared addresses on backup is fine, even if it's redundant, but we should do it at most once. It is fine for me to address in separate issue :+1:

jrainville replied:
I checked the code and the `ProcessBackedupMessages` property is set to true in this condition: `mnemonic != "" && fetchBackup`, so it seems to only happen on re-import. @saledjenic or @igor-sirotin can confirm

osmaczko replied:
I know :sweat_smile:

> I assume, this is only the case when we are recovering account from seedphrase, but even though, it means backup messages processing is enabled for that login session.

jrainville replied:
> it means backup messages processing is enabled for that login session.

Ah right, I missed that part. Indeed, in that case, we probably don't want to send further requests. I'm not sure who to addresses that though. I can open an issue and we can think about it

jrainville replied:
Issue opened here: https://github.com/status-im/status-go/issues/5361

osmaczko commented on protocol/messenger_communities.go:
```diff
@@ -635,6 +658,74 @@ func (m *Messenger) handleCommunityEncryptionKeysRequest(community *communities.
+
+	// Mark Backup as handled to tell the test that we handled the request
+	// This is a bit of a hack since no backup was performed, but it does come from someone's backup
+	state.Response.BackupHandled = true
```
I am opting for not commiting/merging this hack :police_officer:

jrainville replied:
~~ACAB~~ ok officer :upside_down_face:

osmaczko commented on protocol/communities_messenger_token_permissions_test.go:
```diff
@@ -2069,6 +2069,54 @@ func (s *MessengerCommunitiesTokenPermissionsSuite) TestResendEncryptionKeyOnBac
+	s.Require().NoError(err)
+
+	err = s.bob.HandleBackup(s.bob.buildMessageState(), backupMessage, nil)
+	s.Require().NoError(err)
+
+	_, err = WaitOnMessengerResponse(
+		s.owner,
+		func(r *MessengerResponse) bool {
+			return r.BackupHandled
+		},
+		"request for shared accounts not received or handled",
+	)
+	s.Require().NoError(err)
+
+	// Owner will receive the request for addresses and send them back to Bob
+	response, err := WaitOnMessengerResponse(
+		s.bob,
+		func(r *MessengerResponse) bool {
+			return len(r.requestsToJoinCommunity) > 0
+		},
+		"request to join not received",
```
```suggestion
	// Owner will receive the request for addresses and send them back to Bob
	response, err := WaitOnMessengerResponse(
		s.bob,
		func(r *MessengerResponse) bool {
	                 _,_ = s.owner.RetrieveAll()
			return len(r.requestsToJoinCommunity) > 0
		},
		"request to join not received",
	)
	s.Require().NoError(err)
```

This trick should work.

jrainville replied:
Wow! That's genius. It never crossed my mind that I could do this!

jrainville commented:
@osmaczko thanks for the comments. You did solve the issue with the test! It,s way cleaner now. 

Let me know if I should remove the `alreadyHandled` code. I don't have hard opinions.

osmaczko committed:
fix_: filter communities before to avoid passing `alreadyHandled` parameter to `requestCommunityKeysAndSharedAddresses`
